spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/VersionedAppRegistryController.java

```diff
@@ -230,9 +239,41 @@ public void makeDefault(@PathVariable("type") ApplicationType type, @PathVariabl
 	@ResponseStatus(HttpStatus.OK)
 	public void unregister(@PathVariable("type") ApplicationType type, @PathVariable("name") String name,
 			@PathVariable("version") String version) {
+
+		AppRegistration defaultApp = appRegistryService.getDefaultApp(name, type);
+		if (defaultApp != null && defaultApp.getVersion().equals(version)) {
+			String streamWithApp = findStreamContainingAppOf(type, name);
+			if (streamWithApp != null) {
+				throw new AppUsedInStreamException(String.format("The app [%s:%s:%s] you're trying to unregister is " +
+						"already in use on %s stream", name, type, version, streamWithApp));
+			}
+		}
+
 		appRegistryService.delete(name, type, version);
 	}
 
+	private String findStreamContainingAppOf(ApplicationType type, String name) {
+		Iterable<StreamDefinition> streams = streamDefinitionRepository.findAll();
+		for (StreamDefinition stream : streams) {
```

Comment:
I do not think this is the correct approach, checking stream definitions for the `type` and `name`.  Consider this case in simplified syntax

```
app reg time 1.0 default
app reg time 2.0
stream create ticktock --def "time | log"
stream deploy ticktock --properties version.time=2.0
app unreg time 2.0
```
The last line should fail since we should not be unregistering apps that are currently deployed and running.